### PR TITLE
rgw: register daemon in service map with more details

### PR DESCRIPTION
**Pull request by leseb**
From now on, the rgw daemon will register in the service map with
RADOS client ID. This will give us better visibility when
query the Ceph status and also will also us to run multiple rgws with
the same cephX user AND to be displayed individually. Previously, using a
single cephX user to run multiple gateways was possible but the Ceph
status would display a single rgw.
Also, in the same situation it will allow us to gather metrics and
export them through the ceph-mgr Prometheus exporter.

Fixes: https://tracker.ceph.com/issues/49227
Signed-off-by: Sébastien Han <seb@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -1136,6 +1136,8 @@ int RGWRados::register_to_service_map(const string& daemon_type, const map<strin
   if (name.compare(0, 4, "rgw.") == 0) {
     name = name.substr(4);
   }
+  std::string instance_id = stringify(rados->get_instance_id());
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:1139]: what provides stringify?

**Review by mattbenjamin — APPROVED** [review]
lgtm

**Comment by leseb** [comment]
FYI, this could be considered a draft PR, I'm sure they are many places where I'm missing code (like integration tests, unit tests, doc etc), @mattbenjamin feel free to assign someone else more comfortable with that code if necessary :)

**Comment by mattbenjamin** [comment]
Ok.  I'm not sure which of those it needs, but we can take it over, @leseb .  Thanks for kicking this off!

File: src/rgw/rgw_rados.cc (hunk 1 above)
Comment (leseb) [reply src/rgw/rgw_rados.cc:1139]: Not sure, I took this from the rbd-mirror code 😅

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -1136,6 +1136,8 @@ int RGWRados::register_to_service_map(const string& daemon_type, const map<strin
...
+  std::string rgw_service_map_name = name + "." +  metadata["zonegroup_name"] + "." + metadata["zone_name"] + "." + instance_id;
```
Comment (cbodley) [inline src/rgw/rgw_rados.cc:1140]: is this meant to be human-visible anywhere? or are we just adding this stuff to guarantee uniqueness? if the latter, i'd use "zone_id" and drop both zone_name and zonegroup_name

also, none of this accounts for different realms. but that might be out of scope?

Comment (leseb) [reply src/rgw/rgw_rados.cc:1140]: Yes, it's meant to be displayed as part of the ceph status so we want something human-readable. Yes, I wanted to add the realm but it does not seem to be populated anywhere at the moment so I left things the way they were. But if `svc` happens to provide the realm the rgw is part then we can certainly add it!

File: src/rgw/rgw_rados.cc [hunk 3]
```diff
@@ -1136,6 +1136,8 @@ int RGWRados::register_to_service_map(const string& daemon_type, const map<strin
   if (name.compare(0, 4, "rgw.") == 0) {
     name = name.substr(4);
   }
+  std::string instance_id = stringify(rados.get_instance_id());
+  std::string rgw_service_map_name = name + "." +  metadata["zonegroup_name"] + "." + metadata["zone_name"] + "." + instance_id;
   int ret = rados.service_daemon_register(daemon_type, name, metadata);
```
Comment (thotz) [inline src/rgw/rgw_rados.cc:1141]: @leseb The rgw_service_map_name need to passed to this function right?

Comment (leseb) [reply src/rgw/rgw_rados.cc:1141]: > @leseb The `rgw_service_map_name` need to pass to this function right?

Duh! I mixed things up locally 🤦 

> Just out of curiosity, the rbd code passes `name` and `instance_id` to string map of `service_daemon_register() ` but here it is different. Any idea?

I don't know :(

**Comment by sebastian-philipp** [comment]
this need to go through the cephadm suite.

**Comment by leseb** [comment]
> this need to go through the cephadm suite.

How? I don't see any cephadm suite in the list of available Jenkins commands.
Please advise.

File: src/rgw/rgw_rados.cc (hunk 2 above)
Comment (thotz) [reply src/rgw/rgw_rados.cc:1140]: @cbodley : any suggestion for the service_map_name.
1.) when name+realm_name+zone_name
2.) or just name+instance_id sufficient?
3.) adding zone id make it very difficult to understand IMO

**Review by thotz — APPROVED** [review]
LGTM. This change works well with HPA for kubernetes. Tested with rook and minikube

File: src/rgw/rgw_rados.cc (hunk 2 above)
Comment (leseb) [reply src/rgw/rgw_rados.cc]: I added the realm too. Let's test again. Thanks

Comment (leseb) [reply src/rgw/rgw_rados.cc]: Ok, looks like it's not building, if someone can have a look, I'm sure it's easy or someone familiar with the code :) Thanks!

**Comment by mattbenjamin** [comment]
maybe "auto realm = svc.zone->get_realm();" ?

**Review by mattbenjamin — COMMENTED** [review]
lgtm, @cbodley raised the question in standup of whether this is consistent with naming Sage introduced w/cephadm (that does do things like list all the rgws with qualified names)

**Comment by cbodley** [comment]
it looks like `client.rgw.<realmname>.<zonename>` is the format chosen by cephadm

**Comment by leseb** [comment]
> it looks like `client.rgw.<realmname>.<zonename>` is the format chosen by cephadm

But that's the name of the CephX user, right?
In this case, it's going to be confusing since the service map would something like: `rgw.<realmname>.<zonename>.<realmname>.<zonegroupname>.<zonename>.<rados client id>`

**Comment by cbodley** [comment]
then it sounds like @thotz was on the right track here?
> 2.) or just name+instance_id sufficient?

**Comment by leseb** [comment]
> > it looks like `client.rgw.<realmname>.<zonename>` is the format chosen by cephadm
> 
> But that's the name of the CephX user, right?
> In this case, it's going to be confusing since the service map would be something like: `rgw.<realmname>.<zonename>.<realmname>.<zonegroupname>.<zonename>.<rados client id>`

Ok so then maybe, I should revert the realm/zonegroup/zone naming and put that in the CephX client name. I'll keep the RADOS connection ID though.

**Comment by leseb** [comment]
jenkins retest this please

**Comment by harishmunjulur** [comment]
http://pulpito.ceph.com:80/hmunjulu-2021-03-03_05:40:02-rgw-wip-hmunjulu02-testing-2021-03-02-1939-distro-basic-smithi/

https://trello.com/c/DoAG4CDA

**Comment by sebastian-philipp** [comment]
> this need to go through the cephadm suite.

https://pulpito.ceph.com/swagner-2021-03-05_12:08:19-rados:cephadm-wip-swagner3-testing-2021-03-05-1110-distro-basic-smithi/

**Comment by liewegas** [comment]
see #39855

**Pull request merged**